### PR TITLE
feat: major api improvements

### PR DESCRIPTION
**Pull request by GregoryAlbouy**
## Description

Introduction of major changes to the SDK for a more intuitive and convenient usage.

### Usage

<details>
<summary>Minimal configuration: before</summary>

```go
package main

import (
    "context"
    "fmt"

    "github.com/benchttp/engine/runner"
)

func main() {
    // Retrieve a configuration object first
    config := runner.DefaultConfig()
    config.Request = config.Request.WithURL("http://localhost:3000") // tedious, especially with URL being required

    // Instantiate runner and run benchmark
    report, _ := runner. // package name "runner" is too generic
        New(nil). // awkward signature (gets worse with a non nil value)
        Run(context.Background(), config)

    fmt.Println(report.Metrics.ResponseTimes.Mean)
}

```

</details>

<details>
<summary>Minimal configuration: after</summary>

```go
package main

import (
    "context"
    "fmt"

    "github.com/benchttp/sdk/benchttp"
)

func main() {
    report, _ := benchttp.
        DefaultRunner().
        WithNewRequest("GET", "http://localhost:3000", nil).
        Run(context.Background())

    fmt.Println(report.Metrics.ResponseTimes.Mean)
}
```

</details>

<details>
<summary>Advanced configuration: before</summary>

```go
package main

import (
    "context"
    "fmt"
    "net/url"
    "time"

    "github.com/benchttp/engine/runner"
)

func main() {
    payload := []byte(`{"a": "b"}`)
    uri, _ := url.Parse("http://localhost:3000")

    config := runner.Config{
        Request: runner.RequestConfig{
            Method: "POST",
            URL:    uri,
            Body:   runner.NewRequestBody("raw", string(payload)), // 🤢
        },
        Runner: runner.RecorderConfig{ // what is runner.RecorderConfig?
            Requests:       100,
            Concurrency:    10,
            RequestTimeout: 2 * time.Second,
            GlobalTimeout:  30 * time.Second,
        },
    }

    // weird name avoiding naming conflicts
    // weird runner.New signature
    rnr := runner.New(func(progress runner.RecordingProgress) {
        fmt.Println(progress)
    })

    report, _ := rnr.Run(context.Background(), config)

    fmt.Println(report.Metrics.ResponseTimes.Mean)
}

```

</details>

<details>
<summary>Advanced configuration: after</summary>

```go
package main

import (
    "bytes"
    "context"
    "fmt"
    "net/http"
    "time"

    "github.com/benchttp/sdk/benchttp"
)

func main() {
    payload := []byte(`{"a": "b"}`)
    request, _ := http.NewRequest("POST", "http://localhost:3000", bytes.NewReader(payload))

    runner := benchttp.Runner{
        Request:        request,
        Requests:       100,
        Concurrency:    10,
        RequestTimeout: 2 * time.Second,
        GlobalTimeout:  30 * time.Second,
        OnProgress: func(progress benchttp.RecordingProgress) {
            fmt.Println(progress)
        },
    }

    report, _ := runner.Run(context.Background())

    fmt.Println(report.Metrics.ResponseTimes.Mean)
}

```

</details>

### Global changes

- Change module name: `github.com/benchttp/engine` ➡️ `github.com/benchttp/sdk`. It implies to change this repository name to `sdk` after this PR is merged.
- Change package name: `runner` ➡️ `benchttp`: more intuitive naming, such as `runner := benchttp.Runner{}` or `report := benchttp.Report{}` or `errors.Is(err, benchttp.ErrCanceled)`
- Remove Makefile in favor to shell scripts
- Global clean-up: remove irrelevant env file, .gitignore entries, deprecated linters, unused deps

### `runner`

- Remove Config overriding logics. Same results can be achieved by changing signatures from `ToConfig(v interface{}) Config` to `ToConfig(dst *Config)`, which significantly reduces complexity.
- Remove `ConfigFields` types and values: they're became unnecessary after removal of Config overriding logics
- Change `Config.Request` type to `*http.Request`: the runner needs an actual request, not an intermediary representation to be compiled as a request.
- Flatten packages `config`, `report`
- Remove the notion of `Config`, use `Runner` fields instead
- Rename `Runner.onRecordingProgress` to `Runner.OnProgress` (public field)
- Remove superfluous constructor `New`: no internal setup needed, public fields are enough.

### `configparse`

- Change parsing functions signature and implementation so they store the result into a given structure (à la `json.Unmarshal(b []byte, dst interface{})`) . This has two advantages:
  1. Directly overrides the relevant fields of the given `dst`: no need to remember assigned fields then use Config overriding logics
  2. Allows to use any runner configuration as a base, rather than an imposed zero/default value

### CODE REVIEW

File: benchttp/report.go [hunk 1]
```diff
@@ -17,14 +16,14 @@ type Report struct {
 
 // Metadata contains contextual information about a run.
 type Metadata struct {
-	Config        config.Global
+	Config        Runner
```
Comment (moreirathomas) [inline benchttp/report.go:19]:
```suggestion
	Runner        Runner
```

Suggestion: let's go all the way in removing the notion of `Config` and call it `Runner`, aka "produced by this runner".

**Review by moreirathomas — APPROVED** [review]
Very nice, it brings much needed simplicity 👍

Dropping the `Config` struct make the whole thing much lighter.

Single concern I have: now that we configure the `Runner` via its public fields, nothing prevents the user from mutating its values after initialization. I agree that's a stupid thing to do from them, yet doable and I think we can manage an api that prevents this. That's on the todo list I think, you're refactor already brings value that largely outweights this single drawback (for now?).

As it's been quite a long time, I suggest we simply merge, what do you think?

File: benchttp/report.go (hunk 1 above)
Comment (GregoryAlbouy) [reply benchttp/report.go]: Yup good catch!

**Comment by GregoryAlbouy** [comment]
> Single concern I have: now that we configure the Runner via its public fields, nothing prevents the user from mutating its values after initialization. I agree that's a stupid thing to do from them, yet doable and I think we can manage an api that prevents this.

My intuition is that it would be difficult to achieve without losing simplicity (and Go users are used to it, e.g. with `http.Request`), but worth discussing indeed!

**Review by moreirathomas — APPROVED** [review]

**Pull request merged**